dev/tools/lib/archive_publisher.dart

```diff
@@ -0,0 +1,175 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:path/path.dart' as path;
+import 'package:process/process.dart';
+
+class ArchivePublisherException implements Exception {
+  ArchivePublisherException(this.message, [this.result]);
+
+  final String message;
+  final ProcessResult result;
+
+  @override
+  String toString() {
+    String output = 'ArchivePublisherException';
+    if (message != null) {
+      output += ': $message';
+    }
+    final String stderr = result?.stderr ?? '';
+    if (stderr.isNotEmpty) {
+      output += ':\n$result.stderr';
+    }
+    return output;
+  }
+}
+
+/// Publishes the archive created for a particular version and git hash to
+/// the releases directory on cloud storage, and updates the metadata for
+/// releases.
+class ArchivePublisher {
+  ArchivePublisher(
+    this.revision,
+    this.version,
+    this.channel, {
+    this.processManager = const LocalProcessManager(),
+    this.tempDir,
+  });
+
+  /// A git hash describing the revision to publish. It should be the complete
+  /// hash, not just a prefix.
+  final String revision;
+
+  /// A version number for the release (e.g. "1.2.3").
+  final String version;
+
+  /// The channel to publish to. Can be either "dev" or "beta".
+  // TODO(gspencer): support beta: it is currently unimplemented.
+  final String channel;
```

Comment:
should this be an enum if it only supports a limited number of pre-defined values?

Reply:
Done.